lib/cast.js

```diff
@@ -47,6 +49,10 @@ module.exports = function cast(schema, obj, options, context) {
     schema = getSchemaDiscriminatorByValue(schema, obj[schema.options.discriminatorKey]) || schema;
   }
 
+  if (options.upsert && schema.tree._id.required) {
```

Comment:
Also need to check if `obj._id === null`, it's fine to do an upsert on a doc with a required `_id` if `_id` is set or not defined.